### PR TITLE
Create local reflection model: ObjectInfo

### PR DESCRIPTION
**Pull request by dlandi**
1) Removed the individual MethodInfo Broker/Service/Test as it is unnecessary to have multiple Brokers/Services for different parts of DotNet Reflection meta-data. 

2) Added local models mirroring Reflection Metadata we care about, with the root model being ObjectInfo with sub-models rolling up under it.
image

3) Added ObjectInfo Broker and Service and simple Test.

### CODE REVIEW

File: Standard.Reflection.Unit.Tests/Services/Foundations/ObjectInfo/ObjectInfoServiceTests.Logic.Retrieve.cs [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+using Standard.Reflection.Models.Primitives;
+using Standard.Reflection.Services;
+using Standard.Reflection.Unit.Tests.Services.Foundations.Properties;
+using Xunit;
+
+namespace Standard.Reflection.Unit.Tests.Services.Foundations
+{
+    public partial class ObjectInfoServiceTests
+    {
+
```
Comment (cjdutoit) [inline Standard.Reflection.Unit.Tests/Services/Foundations/ObjectInfo/ObjectInfoServiceTests.Logic.Retrieve.cs:19]: Remove empty line

File: Standard.Reflection.Unit.Tests/Services/Foundations/ObjectInfo/ObjectInfoServiceTests.Logic.Retrieve.cs [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+            // when
+            var actualObjectInfo =
+                (ObjectInfo)objsvc.RetrieveObjectInfo(testClass2);
+            var actualObjectInfo3 =
+                (ObjectInfo)objsvc.RetrieveObjectInfo(testClass3);
+
+            // then
+            actualObjectInfo.Should().BeEquivalentTo(expectedObjectInfo);
+            actualObjectInfo3.Should().NotBeEquivalentTo(expectedObjectInfo);
+
```
Comment (cjdutoit) [inline Standard.Reflection.Unit.Tests/Services/Foundations/ObjectInfo/ObjectInfoServiceTests.Logic.Retrieve.cs:41]: Remove empty line

File: Standard.Reflection/Brokers/ObjectInfo/IObjectInfoBroker.cs [hunk 3]
```diff
@@ -0,0 +1,13 @@
+﻿using Standard.Reflection.Models;
```
Comment (cjdutoit) [inline Standard.Reflection/Brokers/ObjectInfo/IObjectInfoBroker.cs:1]: Add copyright

File: Standard.Reflection/Brokers/ObjectInfo/ObjectInfoBroker.cs [hunk 4]
```diff
@@ -0,0 +1,154 @@
+﻿using System;
```
Comment (cjdutoit) [inline Standard.Reflection/Brokers/ObjectInfo/ObjectInfoBroker.cs:1]: Add copyright

File: Standard.Reflection/Brokers/ObjectInfo/ObjectInfoBroker.cs [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using System.Text;
+using Standard.Reflection.Models;
+using Standard.Reflection.Models.Primitives;
+
```
Comment (cjdutoit) [inline Standard.Reflection/Brokers/ObjectInfo/ObjectInfoBroker.cs:8]: Remove empty line

File: Standard.Reflection/Brokers/ObjectInfo/ObjectInfoBroker.cs [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+{
+    public class ObjectInfoBroker : IObjectInfoBroker
+    {
+        public Models.Primitives.IObjectInfo GetObjectInfo(object obj)
+        {
+            var objInfo = new Models.Primitives.ObjectInfo();
+            objInfo.PropInfos = new List<IPropInfo>();
+            objInfo.MethodInfos = new List<IMethodInfo>();
+            objInfo.ImplementedInterfaces = new List<ITypeInfo>();
+
```
Comment (cjdutoit) [inline Standard.Reflection/Brokers/ObjectInfo/ObjectInfoBroker.cs:20]: Remove empty line

File: Standard.Reflection/Brokers/ObjectInfo/ObjectInfoBroker.cs [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+            Type type = obj.GetType();
+
```
Comment (cjdutoit) [inline Standard.Reflection/Brokers/ObjectInfo/ObjectInfoBroker.cs:22]: Remove empty line

File: Standard.Reflection/Brokers/ObjectInfo/ObjectInfoBroker.cs [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+            }
+
+            return objInfo;
+        }
+
+        private ITypeInfo GetIntfcTypeInfo(Type intfcType)
+        {
+            System.Reflection.TypeInfo typeInfo = intfcType.GetTypeInfo();
+            Models.Primitives.TypeInfo modeltypeInfo = new Models.Primitives.TypeInfo();
+
```
Comment (cjdutoit) [inline Standard.Reflection/Brokers/ObjectInfo/ObjectInfoBroker.cs:62]: Remove empty line

File: Standard.Reflection/Brokers/ObjectInfo/ObjectInfoBroker.cs [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+            return modeltypeInfo;
+        }
+
+        private ITypeInfo GetTypeInfo(object obj)
+        {
+            System.Reflection.TypeInfo typeInfo = obj.GetType().GetTypeInfo();
+            Models.Primitives.TypeInfo modeltypeInfo = new Models.Primitives.TypeInfo();
+            modeltypeInfo.CustomAttrs = new List<ITypeInfo>();
+
```
Comment (cjdutoit) [inline Standard.Reflection/Brokers/ObjectInfo/ObjectInfoBroker.cs:81]: Remove empty line

File: Standard.Reflection/Brokers/ObjectInfo/ObjectInfoBroker.cs [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+            propInfo.CanWrite = _propInfo.CanWrite;
+            propInfo.CanRead = _propInfo.CanRead;
+            propInfo.Value = _propInfo.GetValue(obj);
+            propInfo.PropertyType = _propInfo.PropertyType;
+            propInfo.DeclaringType = _propInfo.DeclaringType;
+            propInfo.ReflectedType = _propInfo.ReflectedType;
+
+            propInfo.CustomAttrs = new List<ITypeInfo>();
+            var attrs = _propInfo.GetCustomAttributes(false);
+            foreach (var attr in attrs)
```
Comment (cjdutoit) [inline Standard.Reflection/Brokers/ObjectInfo/ObjectInfoBroker.cs:108]: Add empty line before the foreach

File: Standard.Reflection/Brokers/ObjectInfo/ObjectInfoBroker.cs [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+        {
+            PropInfo propInfo = new PropInfo();
+            propInfo.Name = _propInfo.Name;
+            propInfo.CanWrite = _propInfo.CanWrite;
+            propInfo.CanRead = _propInfo.CanRead;
+            propInfo.Value = _propInfo.GetValue(obj);
+            propInfo.PropertyType = _propInfo.PropertyType;
+            propInfo.DeclaringType = _propInfo.DeclaringType;
+            propInfo.ReflectedType = _propInfo.ReflectedType;
+
```
Comment (cjdutoit) [inline Standard.Reflection/Brokers/ObjectInfo/ObjectInfoBroker.cs:105]: Remove empty line

File: Standard.Reflection/Brokers/ObjectInfo/ObjectInfoBroker.cs [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+            MemberInfo[] myMembers = _methodInfo.DeclaringType.GetMembers();
+
+            for (int i = 0; i < myMembers.Length; i++)
+            {
+                if (myMembers[i].DeclaringType != _methodInfo.DeclaringType ||
+                    myMembers[i].DeclaringType.AssemblyQualifiedName != _methodInfo.DeclaringType.AssemblyQualifiedName)
+                    continue;
+
+                Object[] attrs = myMembers[i].GetCustomAttributes(false);
+                if (attrs.Length > 0)
```
Comment (cjdutoit) [inline Standard.Reflection/Brokers/ObjectInfo/ObjectInfoBroker.cs:132]: Add empty line before the `if`

File: Standard.Reflection/Brokers/ObjectInfo/ObjectInfoBroker.cs [hunk 13]
```diff
@@ -0,0 +1,154 @@
...
+                {
+                    for (int j = 0; j < attrs.Length; j++)
+                    {
+                        Attribute attr = attrs[j] as Attribute;
+                        if (!methodInfo.CustomAttrs.Any(a=>a.Name.Equals(GetTypeInfo(attr).Name)))
```
Comment (cjdutoit) [inline Standard.Reflection/Brokers/ObjectInfo/ObjectInfoBroker.cs:137]: Add empty line before the `if`

File: Standard.Reflection/Brokers/ObjectInfo/ObjectInfoBroker.cs [hunk 14]
```diff
@@ -0,0 +1,154 @@
...
+                        {
+                            methodInfo.CustomAttrs.Add(GetTypeInfo(attr));
+                        }                        
+                    }
+                }
+            }
+
+            return methodInfo;
+        }
+
```
Comment (cjdutoit) [inline Standard.Reflection/Brokers/ObjectInfo/ObjectInfoBroker.cs:147]: Remove empty lines

File: Standard.Reflection/Brokers/ObjectInfo/ObjectInfoBroker.cs [hunk 15]
```diff
@@ -0,0 +1,154 @@
...
+
+    }
+
```
Comment (cjdutoit) [inline Standard.Reflection/Brokers/ObjectInfo/ObjectInfoBroker.cs:150]: Remove empty line

File: Standard.Reflection/Models/Primitives/IMethodInfo.cs [hunk 16]
```diff
@@ -0,0 +1,12 @@
+﻿using System;
```
Comment (cjdutoit) [inline Standard.Reflection/Models/Primitives/IMethodInfo.cs:1]: Add copyright

File: Standard.Reflection/Models/Primitives/IObjectInfo.cs [hunk 17]
```diff
@@ -0,0 +1,12 @@
+﻿using System.Collections.Generic;
```
Comment (cjdutoit) [inline Standard.Reflection/Models/Primitives/IObjectInfo.cs:1]: Add copyright

File: Standard.Reflection/Models/Primitives/IPropInfo.cs [hunk 18]
```diff
@@ -0,0 +1,16 @@
+﻿using System;
```
Comment (cjdutoit) [inline Standard.Reflection/Models/Primitives/IPropInfo.cs:1]: Add copyright

File: Standard.Reflection/Models/Primitives/ITypeInfo.cs [hunk 19]
```diff
@@ -0,0 +1,20 @@
+﻿using System;
```
Comment (cjdutoit) [inline Standard.Reflection/Models/Primitives/ITypeInfo.cs:1]: Add copyright

File: Standard.Reflection/Models/Primitives/MethodInfo.cs [hunk 20]
```diff
@@ -0,0 +1,20 @@
+﻿using System;
```
Comment (cjdutoit) [inline Standard.Reflection/Models/Primitives/MethodInfo.cs:1]: Add copyright

File: Standard.Reflection/Models/Primitives/MethodInfo.cs [hunk 21]
```diff
@@ -0,0 +1,20 @@
...
+using System.Collections.Generic;
+using System.Globalization;
+using System.Reflection;
+using System.Text;
+
+namespace Standard.Reflection.Models.Primitives
+{
+    public class MethodInfo : IMethodInfo
+    {
+
```
Comment (cjdutoit) [inline Standard.Reflection/Models/Primitives/MethodInfo.cs:11]: Remove all the empty lines

File: Standard.Reflection/Models/Primitives/ObjectInfo.cs [hunk 22]
```diff
@@ -0,0 +1,14 @@
+﻿using System;
```
Comment (cjdutoit) [inline Standard.Reflection/Models/Primitives/ObjectInfo.cs:1]: Add copyright

File: Standard.Reflection/Models/Primitives/PropInfo.cs [hunk 23]
```diff
@@ -0,0 +1,28 @@
+﻿using Standard.Reflection.Models.Primitives;
```
Comment (cjdutoit) [inline Standard.Reflection/Models/Primitives/PropInfo.cs:1]: Add copyright

File: Standard.Reflection/Models/Primitives/TypeInfo.cs [hunk 24]
```diff
@@ -0,0 +1,31 @@
+﻿using System;
```
Comment (cjdutoit) [inline Standard.Reflection/Models/Primitives/TypeInfo.cs:1]: Add copyright

File: Standard.Reflection/Services/Foundations/ObjectInfo/ObjectInfoService.Exceptions.cs [hunk 25]
```diff
@@ -0,0 +1,33 @@
...
+        private IObjectInfo TryCatch(ReturningObjectInfoFunction returningObjectInfoFunction)
+        {
+            try
+            {
+                return returningObjectInfoFunction();
+            }
+            catch (Exception exception)
+            {
+                var failedObjectInfoServiceException = new ObjectInfoServiceException((Xeptions.Xeption)exception);
+
```
Comment (cjdutoit) [inline Standard.Reflection/Services/Foundations/ObjectInfo/ObjectInfoService.Exceptions.cs:26]: Remove empty line

File: Standard.Reflection.Unit.Tests/Services/Foundations/ObjectInfo/ObjectInfoServiceTests.cs [hunk 26]
```diff
@@ -0,0 +1,56 @@
...
+        {
+            this.objectInfoBrokerMock = new Mock<IObjectInfoBroker>();
+            this.objectInfoService = new ObjectInfoService(this.objectInfoBrokerMock.Object);
+        }
+
+        public class IsStandardCompliant : Attribute
+        {
+            public int LevelOfCompliance { get; set; }
+        }
+
```
Comment (cjdutoit) [inline Standard.Reflection.Unit.Tests/Services/Foundations/ObjectInfo/ObjectInfoServiceTests.cs:36]: Remove empty line

**Comment by hassanhabib** [comment]
This PR should be rejected. It goes against The Standard rules of file/folder organization.

**Comment by dlandi** [comment]
> This PR should be rejected. It goes against The Standard rules of file/folder organization.

Hi @hassanhabib .  I need more info. Do you mean this sort of organization?
image

**Comment by hassanhabib** [comment]
> > This PR should be rejected. It goes against The Standard rules of file/folder organization.
> 
> Hi @hassanhabib . I need more info. Do you mean this sort of organization? image

First off - your brokers implementation have logic in them - that's `if statements`, `foreach` ... etc
Secondly - `Primitives` you seem to be combining multiple models in the same folder. Folders represent `Models`, `Services`, Brokers` and `Exposers` of one entity
Thirdly - commit history here is not pure. doesn't follow any of our commit conventions

I could go deeper but off the bat - this PR is a no go.

**Comment by dlandi** [comment]
I'll just post this an entirely separate repo, and not associate with The Standard. The first item is a deal breaker for me.  The rest of the comments are just book-keeping. Sorry, it didn't work out.

**Pull request closed without merging**